File: examples/python/best-practices/separate-business-logic.py
# business_logic.py
@durable_step
def validate_order(step_context: StepContext, order: dict) -> dict:
    if not order.get("items"):
        raise ValueError("Order must have items")
    return {**order, "validated": True}

# handler.py
@durable_execution
def lambda_handler(event: dict, context: DurableContext) -> dict:
    order = event["order"]
    validated_order = context.step(validate_order(order))
    return {"status": "completed", "order_id": validated_order["order_id"]}

File: examples/java/best-practices/separate-business-logic.java
// BusinessLogic.java
static Map<String, Object> validateOrder(Map<String, Object> order) {
    if (!order.containsKey("items") || ((List<?>) order.get("items")).isEmpty()) {
        throw new IllegalArgumentException("Order must have items");
    }
    var result = new HashMap<>(order);
    result.put("validated", true);
    return result;
}

// Handler.java
var order = (Map<String, Object>) input.get("order");
var validatedOrder = ctx.step("validate_order", Map.class,
    stepCtx -> validateOrder(order));
return Map.of("status", "completed", "order_id", validatedOrder.get("order_id"));
